src/dotnet/ShellShimMaker/EnvironmentPathFactory.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Xml.Xsl;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.Configurer;
+using Microsoft.Extensions.EnvironmentAbstractions;
+
+namespace Microsoft.DotNet.ShellShimMaker
+{
+    internal static class EnvironmentPathFactory
+    {
+        public static IEnvironmentPath CreateEnvironmentPath(
+            CliFolderPathCalculator cliFolderPathCalculator = null,
+            bool hasSuperUserAccess = false,
+            IEnvironmentProvider environmentProvider = null)
+        {
+            if (cliFolderPathCalculator == null)
+            {
+                cliFolderPathCalculator = new CliFolderPathCalculator();
+            }
+
+            if (environmentProvider == null)
+            {
+                environmentProvider = new EnvironmentProvider();
+            }
+
+            IEnvironmentPath environmentPath = new DoNothingEnvironmentPath();
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                environmentPath = new WindowsEnvironmentPath(
+                    cliFolderPathCalculator.ExecutablePackagesPath,
+                    Reporter.Output,
+                    environmentProvider);
+            }
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux) && hasSuperUserAccess)
```

Comment:
With the code as is here for linux and OSX, are users ever going to see the message to edit the paths? It seems like the Linux and OSX Environment paths will only be created when hasSuperUserAccess is true, in which case, adding to the path will work.

While in the case where there is no superUserAccess, they will just get the DoNothingEnvironmentPath.

Reply:
> While in the case where there is no superUserAccess, they will just get the DoNothingEnvironmentPath.

This object will be created again in InstallCommand(problem of partial code review) and it will have the interface without the addToPath method.

As for why this object will need to be created twice. The interface of resolving build-in command is Func<T> with args[]. There is no place to pass down object

Reply:
But I am not very satisfied with the way I handling sudo/no sudo - when in sudo you can call add to path. Currently it requires flags and a big factory method. But I cannot find a better way.